src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+	msleep(200);
+
+	memset(&report->header, 0, sizeof(report->header));
+	report->header.write.report_id = 0x4;
+	report->checksum = gskill_calculate_checksum((uint8_t*)report,
+						     sizeof(*report));
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GET_SET_MACRO,
+				       (uint8_t*)report, sizeof(*report),
+				       HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to write macro for profile %d button %d to mouse: %d\n",
+			  profile, button, rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+static int
+gskill_probe(struct ratbag_device *device)
+{
+	struct gskill_data *drv_data = NULL;
+	struct ratbag_profile *profile;
+	unsigned int active_idx;
+	int ret;
+
+	ret = ratbag_open_hidraw(device);
+	if (ret)
+		return ret;
+
+	drv_data = zalloc(sizeof(*drv_data));
+	ratbag_set_drv_data(device, drv_data);
+
+	ret = gskill_get_profile_count(device);
+	if (ret < 0)
+		goto err;
+
+	/*
+	 * TODO: Add proper support for enabling/disabling profiles in
+	 * libratbag. For now we workaround this by just setting the profile
+	 * count to 5
+	 */
+	if (ret < GSKILL_PROFILE_MAX) {
+		log_info(device->ratbag,
+			 "We don't support dynamically enabling/disabling profiles yet, sorry! Setting profile count of mouse to 5\n");
+		ret = gskill_set_profile_count(device, GSKILL_PROFILE_MAX);
+		if (ret < 0)
+			goto err;
+	}
+
+	ret = gskill_get_firmware_version(device);
+	if (ret < 0)
+		goto err;
+
+	log_debug(device->ratbag,
+		 "Firmware version: %d\n", ret);
+
+	ratbag_device_init_profiles(device, GSKILL_PROFILE_MAX, GSKILL_NUM_DPI,
+				    GSKILL_BUTTON_MAX);
+
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_QUERY_CONFIGURATION);
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_RESET_PROFILE);
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_BUTTON_KEY);
+	ratbag_device_set_capability(device, RATBAG_DEVICE_CAP_BUTTON_MACROS);
+
+	ret = gskill_get_active_profile_idx(device);
+	if (ret < 0)
+		goto err;
+
+	active_idx = ret;
+	list_for_each(profile, &device->profiles, link) {
+		if (profile->index == active_idx) {
+			profile->is_active = true;
+			break;
+		}
+	}
+
+	return 0;
+
+err:
+	if (drv_data) {
+		ratbag_set_drv_data(device, NULL);
+		free(drv_data);
+	}
+
+	ratbag_close_hidraw(device);
+	return ret;
+}
+
+static void
+gskill_read_profile(struct ratbag_profile *profile, unsigned int index)
+{
+	struct ratbag_device *device = profile->device;
+	struct gskill_data *drv_data = ratbag_get_drv_data(device);
+	struct gskill_profile_data *pdata = &drv_data->profile_data[index];
+	struct gskill_profile_report *report = &pdata->report;
+	uint8_t checksum;
+	int rc, retries;
+
+	/*
+	 * There's a couple of situations where after various commands, the
+	 * mouse will get confused and send the wrong profile. Keep trying
+	 * until we get what we want.
+	 *
+	 * As well, getting the wrong profile is sometimes a sign from the
+	 * mouse we're doing something wrong.
```

Comment:
Yep, that should hopefully not happen. I wonder if this should stay in the final code. Or is this happening even now that you are actually doing the right commands?